crates/ef-testing/src/evm_sequencer/setup.rs

```diff
@@ -0,0 +1,246 @@
+use blockifier::abi::abi_utils::{
+    get_erc20_balance_var_addresses, get_storage_var_address, get_uint256_storage_var_addresses,
+};
+use blockifier::state::state_api::{State, StateResult};
+use reth_primitives::{Address, Bytes};
+use revm_primitives::U256;
+use starknet_api::core::Nonce;
+use starknet_api::hash::StarkFelt;
+use starknet_api::StarknetApiError;
+
+use super::constants::{
+    CONTRACT_ACCOUNT_CLASS_HASH, EOA_CLASS_HASH, FEE_TOKEN_ADDRESS, KAKAROT_ADDRESS,
+    PROXY_CLASS_HASH,
+};
+use super::types::FeltSequencer;
+use super::utils::{
+    class_hash_to_starkfelt, compute_starknet_address, contract_address_to_starkfelt,
+    split_bytecode_to_starkfelt, split_u256,
+};
+use super::KakarotSequencer;
+
+pub trait InitializeEvmState {
+    fn initialize_contract(
+        &mut self,
+        evm_address: &Address,
+        bytecode: &Bytes,
+        nonce: U256,
+        storage: Vec<(U256, U256)>,
+    ) -> StateResult<()>;
+
+    fn fund(&mut self, evm_address: &Address, balance: U256) -> StateResult<()>;
+}
+
+impl InitializeEvmState for KakarotSequencer {
+    fn initialize_contract(
+        &mut self,
+        evm_address: &Address,
+        bytecode: &Bytes,
+        nonce: U256,
+        evm_storage: Vec<(U256, U256)>,
+    ) -> StateResult<()> {
+        let nonce = StarkFelt::from(TryInto::<u128>::try_into(nonce).map_err(|err| {
+            StarknetApiError::OutOfRange {
+                string: err.to_string(),
+            }
+        })?);
+        let starknet_address = compute_starknet_address(evm_address);
+        let evm_address = Into::<FeltSequencer>::into(*evm_address).into();
+
+        let mut storage = vec![
+            (
+                get_storage_var_address("evm_address", &[]).unwrap(), // safe unwrap: var is ASCII
+                evm_address,
+            ),
+            (
+                get_storage_var_address("is_initialized_", &[]).unwrap(), // safe unwrap: var is ASCII
+                StarkFelt::from(1u8),
+            ),
+            (
+                get_storage_var_address("Ownable_owner", &[]).unwrap(), // safe unwrap: var is ASCII
+                contract_address_to_starkfelt(&KAKAROT_ADDRESS),
+            ),
+            (
+                get_storage_var_address("bytecode_len_", &[]).unwrap(), // safe unwrap: var is ASCII
+                StarkFelt::from(bytecode.len() as u32),
+            ),
+            (
+                get_storage_var_address("kakarot_address", &[]).unwrap(), // safe unwrap: var is ASCII
+                contract_address_to_starkfelt(&KAKAROT_ADDRESS),
+            ),
+            (
+                get_storage_var_address("nonce", &[]).unwrap(), // safe unwrap: var is ASCII
+                nonce,
+            ),
+            (
+                get_storage_var_address("_implementation", &[]).unwrap(), // safe unwrap: var is ASCII
+                if bytecode.is_empty() && evm_storage.is_empty() {
```

Comment:
here I would set the nonce i the if/else as well. There are all the shared things, and one if/else for everything EOA vs CA

Reply:
there are quite a lot of things that aren't shared, the reason I did it that way is that setting one variable in the storage doesn't change much imo and makes the code easier to read.
The differences between eoa and ca are:
- eoa: kakarot_address
- ca: bytecode, bytecode_len_, storage_, nonce 

The similarities are: evm_address, is_initialized_, _implementation
To note is that since bytecode and storage will be empty for eoa, the only values being set that aren't needed are nonce and bytecode_len_